webextension/privileged/ui/CloudDownloadsView.jsm

```diff
@@ -0,0 +1,760 @@
+      this.selectedProvider.value.typeSpecificData.default);
+
+    // create download directory if it doesn't exist
+    try {
+      await OS.File.makeDir(providerDownloadFolder, {ignoreExisting: true});
+    } catch (err) {
+      Cu.reportError(err);
+      return null;
+    }
+    return providerDownloadFolder;
+  },
+
+  async handleMove(download) {
+    const providerDownloadFolder = await this.checkProviderDownloadFolder();
+    if (download.succeeded) {
+      await this._moveDownload(download, providerDownloadFolder);
+    }
+  },
+
+  async handleLocalMove(download) {
+    const downloadsDir = await Downloads.getPreferredDownloadsDirectory();
+    const downloadsDirExists = await this.checkIfAssetExists(downloadsDir);
+    if (downloadsDirExists) {
+      if (download.succeeded) {
+        await this._moveDownload(download, downloadsDir);
+      }
+    }
+  },
+};
+
+
+var WindowListener = {
+  setupBrowserUI: function wm_setupBrowserUI(window) {
+    const utils = window.QueryInterface(Ci.nsIInterfaceRequestor).getInterface(Ci.nsIDOMWindowUtils);
+    utils.loadSheetUsingURIString(CloudDownloadsView.stylesURL, Ci.nsIDOMWindowUtils.AUTHOR_SHEET);
+  },
+
+  tearDownBrowserUI: function wm_tearDownBrowserUI(window) {
+    const utils = window.QueryInterface(Ci.nsIInterfaceRequestor).getInterface(Ci.nsIDOMWindowUtils);
+    utils.removeSheetUsingURIString(CloudDownloadsView.stylesURL, Ci.nsIDOMWindowUtils.AUTHOR_SHEET);
+  },
+
+  // nsIWindowMediatorListener functions
+  onOpenWindow: function wm_onOpenWindow(xulWindow) {
+    // A new window has opened
+    const domWindow = xulWindow.QueryInterface(Ci.nsIInterfaceRequestor)
+      .getInterface(Ci.nsIDOMWindow);
+
+    // Wait for it to finish loading
+    domWindow.addEventListener("load", async function listener() {
+      // If this is a browser window or places library window then setup its UI
+      if (domWindow.document.documentElement.getAttribute("windowtype") === "navigator:browser" ||
+          domWindow.document.documentElement.getAttribute("windowtype") === "Places:Organizer" ) {
+        if (!CloudDownloadsView.stylesURL) {
+          return;
+        }
+        WindowListener.setupBrowserUI(domWindow);
+
+        const moveDownloadMenuItem = domWindow.document.getElementById("moveDownload");
+        if (moveDownloadMenuItem) {
+          return;
+        }
+        await CloudDownloadsView.registerContextMenu(domWindow);
+      }
+    }, {once: true});
+  },
+};
+
+/**
+ * generic pref that shows if cloud storage API is in use, by default set to false.
+ */
+XPCOMUtils.defineLazyPreferenceGetter(CloudDownloadsView, "gIsAPIEnabled",
+  CLOUD_SERVICES_PREF + "api.enabled", false, () => CloudDownloadsView.toggleAPIEnabledState());
+
+XPCOMUtils.defineLazyPreferenceGetter(CloudDownloadsInternal, "downloadDirSetting",
+  "browser.download.dir", "", () => CloudDownloadsInternal._setDefaultDownloadDirIconURL());
```

Comment:
Looks like this should also watch `browser.download.folderList`, right?

Reply:
we don't need to watch browser.download.folderList as we need to know the download paths set for either of the folderList values 0 - Desktop, 1 -Downloads , 2 - Custom Path, all these respective download path values when changed are stored in browser.download.dir

The only exception is default profile when browser.download.dir is not set (browsre.download.folderList as 1), this should be handled when we first invoke setDefaultDownloadDirIconURL in registerContextMenu on initialization